numpy/core/src/multiarray/item_selection.c

```diff
@@ -2202,7 +2206,164 @@ count_nonzero_bytes(const npy_uint8 *d, npy_uintp unrollx)
     return unrollx - zero_count;
 }
 
+
+static NPY_INLINE NPY_GCC_OPT_3 npy_uintp
+count_nonzero_int16_simd(npy_int16 *d, npy_uintp unrollx)
+{
+    npy_uintp zero_count = 0;
+    const npy_uintp innerloop_jump = NPY_MAX_UINT16;
+    const npy_int16 *end = d + unrollx;
+
+    const npyv_u16 vone = npyv_setall_u16(1); 
+    const npyv_u16 vzero = npyv_zero_u16();   
+
+    npy_int16 *target = d;
+    while (d<end) {
+        npyv_u16 vsum16 = npyv_zero_u16(); 
+        target = PyArray_MIN(target+innerloop_jump, end);
+        for (; d<target; d+=npyv_nlanes_u16) {
+            npyv_u16 vt = npyv_cvt_u16_b16(npyv_cmpeq_u16(npyv_load_u16(d), vzero)); 
+            vt = npyv_and_u16(vt, vone);
+            vsum16 = npyv_add_u16(vsum16, vt);
+        }
+
+        zero_count += npyv_sumup_u16(vsum16);
+    }
+
+    return unrollx - zero_count;
+}
+
+
+static NPY_INLINE NPY_GCC_OPT_3 npy_uintp
+count_nonzero_int32_simd(npy_int32 *d, npy_uintp unrollx)
+{
+    npy_uintp zero_count = 0;
+    const npy_uintp innerloop_jump = NPY_MAX_UINT32;
+    const npy_int32 *end = d + unrollx;
+
+    const npyv_u32 vone = npyv_setall_u32(1); 
+    const npyv_u32 vzero = npyv_zero_u32();   
+
+    npy_int32 *target = d;
+    while (d<end) {
+        npyv_u32 vsum32 = npyv_zero_u32(); 
+        target = PyArray_MIN(target+innerloop_jump, end);
+        for (; d<target; d+=npyv_nlanes_u32) {
+            npyv_u32 vt = npyv_cvt_u32_b32(npyv_cmpeq_u32(npyv_load_u32(d), vzero)); 
+            vt = npyv_and_u32(vt, vone);
+            vsum32 = npyv_add_u32(vsum32, vt);
+        }
+        zero_count += npyv_sum_u32(vsum32);    
```

Comment:
since the inner loop guard against wrapping around zero(`assuming allocated array will exceed (max_uint*4)b`!!) then the summation should also follow the same approach. you can zero-extend the odd and even elements then perform sum64 instead of implementing `npyv_sumup_u32`.
```suggestion
       const npyv_u32 maskevn = npyv_reinterpret_u32_u64(npyv_setall_u64(0xffffffffULL));
        npyv_u64 odd  = npyv_shri_u64(npyv_reinterpret_u64_u32(vsum32), 32);
        npyv_u64 even = npyv_reinterpret_u64_u32(npyv_and_u32(vsum32, maskevn));
        zero_count   += npyv_sum_u64(npyv_add_u64(odd, even));    
```

Reply:
Is there any advantage to using the suggested approach? This looks like slower.

Reply:
> Is there any advantage to using the suggested approach?

to avoid the possibility of overflow.

> This looks like slower.

Not at all. almost the same cost of `npyv_sum_u32`.
But anyway it will be called once every `NPY_MAX_UINT32`!. 

for a better approach, you can just drop the inner loop and compute it directly same as you do with u64 but in that case, you will need to handle any scalars above `NPY_MAX_UINT32` via pure `C`.